[feat] bash support

CruizeMissile opened the pull request:
I work on both linux and windows. As it is a pain to get zsh working on some terminal emulators I added support for fw and workon to be sourced in bash as well.

For autocomplete, clap has the ability to generate a completion script so I used that to create the completion part of the setup script. Please let me know if I forgot anything.

Thanks

CruizeMissile commented:
After extended use of the clap generated completion I feel like it is lacking project completion. The clap completion works fine for the commands and options that can be given but has no idea about positional project completion. Going to write an actual bash completion instead

CruizeMissile commented:
Added back completion to the startup script for bash. Try it out and let me know.

bomgar commented:
There are more commands that require project completion:
```
            remove) COMPREPLY=( $( compgen -W "$(__fw_projects)" -- "$cur" ) ) ;;
            inspect) COMPREPLY=( $( compgen -W "$(__fw_projects)" -- "$cur" ) ) ;;
            add-remote) COMPREPLY=( $( compgen -W "$(__fw_projects)" -- "$cur" ) ) ;;
            remove-remote) COMPREPLY=( $( compgen -W "$(__fw_projects)" -- "$cur" ) ) ;;
            print-path) COMPREPLY=( $( compgen -W "$(__fw_projects)" -- "$cur" ) ) ;;
            update) COMPREPLY=( $( compgen -W "$(__fw_projects)" -- "$cur" ) ) ;;
            export-project) COMPREPLY=( $( compgen -W "$(__fw_projects)" -- "$cur" ) ) ;;
```

Would you mind adding the bash setup to the documentation?

I tried to use the completion for `sync` but the options like `--no-ff-merge` did not work.

Could you rebase your branch? There are some commits here that are already on master.

CruizeMissile commented:
@bomgar I rewrote the bash completion, rebased the branch and updated the install documentation to add bash as well. (Unsure what file format .org is. Guessing it is emacs org mode that github can understand not sure on formatting)

For the completion I based it off of git's completion script. This made it more verbose but also more customization for each command. Commands are not project and tag aware when it comes to completion.

Try it out and let me know if there are any issues.

Thanks

bomgar commented:
I doesn't work. 

When i press tab:
```
 fw bash: _get_comp_words_by_ref: command not found
bash: [: 1: unary operator expected
```

bomgar commented:
We use travis to verify the syntax https://github.com/brocode/fw/blob/master/.travis.yml#L36 for our zsh integration. I think you should do the same for the bash script.

bomgar commented:
```

bomgar commented:
```
» bash --version
GNU bash, version 5.0.0(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

CruizeMissile commented:
@bomgar `_get_comp_words_by_ref` is the package bash-completion. `git` and `cargo` depend on this package in order to have completion.

I have taken the related functions from bash-completion and added it to the completion script.

As for the command `eval $(cargo run -- print-bash-setup)`  `eval` takes a string to evaluate so the correct command would be `eval  "$(cargo run -- print-bash-setup)"`

bomgar commented:
Merged. 

btw you should configure your editor to remove trailing spaces
